FreeRTOS/coreJSON
# Add JSON Unit Tests

This PR adds the following [unit tests from PR #1139 of C SDK Development Branch](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1172)

* Add skeleton for json unit tests

* Add json_utest to coverage target

* Add basic happy path test

* Add CMakeLists.txt for JSON library

* Do not include null term char

* Add some simple cases

* Add literals and exponents and collection test cases

* Add cases for trailing space

* Complete coverage for shortestUTF8

* Complete coverage for skipAnyScalar and its callee methods

* Complete coverage for skipExponent

* Complete coverage for skipArrayScalars

* Complete coverage for skipObjectScalars

* Complete coverage for nextKeyValuePair

* Complete coverage for max depth cases

* Make all single string scalars into complete JSON document with key and value when possible

* Make JSON_Search look up the same cases as JSON_Validate

* Ensure that all cases tried in JSON_Validate are also tried in JSON_Search

* Change unreachable branches to asserts

* Add lexicon.txt

* Update lexicon.txt

* Update include direcories to have only what's needed for json library and tests

* Remove cmake_minimum_required for json/utest/CMakeLists.txt

* Address PR comments

* Update lexicon.txt

* Use json library for CBMC PR

* Add invalid params tests for json unit tests

* Complete coverage for search

* Add JSON_Search Bad and NULL parameter tests

* Complete coverage for JSON library

* Reflect JSON_Search invalid documents cases to reflect JSON_Validate

* Update lexicon

* Remove json library that is private to shadow demo
